File: backend/internal/domain/personnel.go
package domain

import "time"

// Address คือที่อยู่แบบแยกฟิลด์ตามมาตรฐานไทย (ใช้ร่วมกับ students/guardians ภายหลังได้)
type Address struct {
	HouseNo     string
	Moo         string
	Road        string
	Subdistrict string
	District    string
	Province    string
	PostalCode  string
}

// PersonnelProfile คือข้อมูลโปรไฟล์ที่ไม่อ่อนไหวสูง (แก้ไขได้)
type PersonnelProfile struct {
	Prefix    string
	FirstName string
	LastName  string
	BirthDate *time.Time
	Phone     string
	Email     string
	Address   Address
}

// Personnel คือ entity ระดับ DB ของบุคลากร — national_id เก็บเป็น ciphertext เท่านั้น
// (service เป็นผู้ถอดรหัส/mask; ห้ามให้ repository รู้ plaintext)
type Personnel struct {
	ID                 string
	SchoolID           string
	UserID             string
	NationalIDEnc      []byte
	NationalIDHash     string
	CivilServantIDEnc  []byte // nil = ไม่มี
	CivilServantIDHash string
	Profile            PersonnelProfile
	PhotoPath          string

	// fields ที่ join มาจากตาราง users
	Username string
	Role     string
	IsActive bool

	CreatedAt time.Time
	UpdatedAt time.Time
}

// NewPersonnel คือ payload สำหรับสร้างบุคลากร (พร้อมบัญชี user) — ค่าอ่อนไหวถูกเข้ารหัสมาแล้วโดย service
type NewPersonnel struct {
	// บัญชี user (สำหรับ login)
	Username      string
	PasswordHash  string
	Role          string
	IsSchoolAdmin bool

	// ข้อมูลอ่อนไหว (เข้ารหัส + hash มาแล้ว)
	NationalIDEnc      []byte
	NationalIDHash     string
	CivilServantIDEnc  []byte
	CivilServantIDHash string

	Profile   PersonnelProfile
	PhotoPath string
}

// UpdatePersonnel คือ payload แก้ไขบุคลากร — เปลี่ยนเลขบัตรเฉพาะเมื่อ ChangeNationalID/ChangeCivilID = true
type UpdatePersonnel struct {
	Profile PersonnelProfile

	ChangeNationalID bool
	NationalIDEnc    []byte
	NationalIDHash   string

	ChangeCivilID      bool
	CivilServantIDEnc  []byte
	CivilServantIDHash string
}

// AuditEntry คือรายการ audit สำหรับการเข้าถึงข้อมูลส่วนบุคคล (PDPA)
// Detail เก็บเฉพาะชื่อ field ที่แตะ ไม่เก็บค่าจริงของข้อมูลอ่อนไหว
type AuditEntry struct {
	SchoolID    string
	ActorUserID string
	Action      string // view, create, update, delete, export
	TargetType  string // personnel, student, ...
	TargetID    string
	Detail      map[string]any
	IPAddress   string
}

// Audit actions
const (
	AuditView   = "view"
	AuditCreate = "create"
	AuditUpdate = "update"
	AuditDelete = "delete"
	AuditExport = "export"
)

File: frontend/src/shared/schemas/personnel.ts
import { z } from "zod";

// schema ของ personnel — ต้องตรงกับ backend (service.PersonnelListItem / PersonnelDetail)
// เลขบัตรประชาชนมาเป็น masked เท่านั้น (PDPA) ไม่มีเลขเต็มใน response

// ตำแหน่งบุคลากร: ครู/ผู้บริหาร (subset ของ user role)
export const personnelRoleSchema = z.enum(["teacher", "executive"]);
export type PersonnelRole = z.infer<typeof personnelRoleSchema>;

export const personnelRoleLabel: Record<PersonnelRole, string> = {
  teacher: "ครู",
  executive: "ผู้บริหาร",
};

export const addressSchema = z.object({
  house_no: z.string(),
  moo: z.string(),
  road: z.string(),
  subdistrict: z.string(),
  district: z.string(),
  province: z.string(),
  postal_code: z.string(),
});
export type AddressData = z.infer<typeof addressSchema>;

// รายการย่อสำหรับตาราง
export const personnelListItemSchema = z.object({
  id: z.string(),
  user_id: z.string(),
  username: z.string(),
  role: z.string(),
  is_active: z.boolean(),
  prefix: z.string(),
  first_name: z.string(),
  last_name: z.string(),
  national_id_masked: z.string(),
  phone: z.string(),
  created_at: z.string(),
});
export type PersonnelListItem = z.infer<typeof personnelListItemSchema>;

export const personnelListSchema = z.array(personnelListItemSchema);

// รายละเอียดเต็ม (เลขบัตรยัง masked)
export const personnelDetailSchema = z.object({
  id: z.string(),
  user_id: z.string(),
  username: z.string(),
  role: z.string(),
  is_active: z.boolean(),
  prefix: z.string(),
  first_name: z.string(),
  last_name: z.string(),
  national_id_masked: z.string(),
  civil_servant_id_masked: z.string(),
  birth_date: z.string(),
  phone: z.string(),
  email: z.string(),
  address: addressSchema,
  photo_path: z.string(),
  created_at: z.string(),
  updated_at: z.string(),
});
export type PersonnelDetail = z.infer<typeof personnelDetailSchema>;

// response ของการสร้าง (คืน id)
export const createdIdSchema = z.object({ id: z.string() });

// response ข้อความ (update/delete)
export const messageSchema = z.object({ message: z.string() });

// payload ที่ส่งไป backend (snake_case ตาม JSON contract)
export type PersonnelAddressBody = AddressData;

export type CreatePersonnelBody = {
  username: string;
  password: string;
  role: PersonnelRole;
  national_id: string;
  civil_servant_id: string;
  prefix: string;
  first_name: string;
  last_name: string;
  birth_date: string;
  phone: string;
  email: string;
  address: PersonnelAddressBody;
};

export type UpdatePersonnelBody = {
  national_id: string;
  civil_servant_id: string;
  prefix: string;
  first_name: string;
  last_name: string;
  birth_date: string;
  phone: string;
  email: string;
  address: PersonnelAddressBody;
};
